src/simulator/diff_drive_model.cpp

```diff
@@ -0,0 +1,188 @@
+    odometry_x = 0.0;
+    odometry_y = 0.0;
+    odometry_w = 0.0;
+    target_linear_vel = 0.0;
+    target_angular_vel = 0.0;
+    pose_.angle = 0;
+    odom_msg.header.frame_id = "/odom";
+    odom_msg.child_frame_id = "/base_link";
+
+}
+
+void DiffDriveModel::Step(const double &dt) {
+    ros::Time current_time = ros::Time::now();
+    // Update the linear velocity based on the linear acceleration limits
+	if (linear_vel < target_linear_vel) {
+		// Must increase linear speed
+		if (CONFIG_linear_pos_accel_limit == 0.0 || target_linear_vel - linear_vel < CONFIG_linear_pos_accel_limit)
+			linear_vel = target_linear_vel;
+		else
+			 linear_vel += CONFIG_linear_pos_accel_limit; 
+	} else if (linear_vel > target_linear_vel) {
+		// Must decrease linear speed
+		if (CONFIG_linear_neg_accel_limit == 0.0 || linear_vel - target_linear_vel < CONFIG_linear_neg_accel_limit)
+			linear_vel = target_linear_vel;
+		else
+			 linear_vel -= CONFIG_linear_neg_accel_limit; 
+	}
+
+	// Update the angular velocity based on the angular acceleration limits
+	if (angular_vel < target_angular_vel) {
+		// Must increase angular speed
+		if (CONFIG_angular_pos_accel_limit == 0.0
+			|| target_angular_vel - angular_vel < CONFIG_angular_pos_accel_limit)
+			angular_vel = target_angular_vel;
+		else
+			 angular_vel += CONFIG_angular_pos_accel_limit; 
+	} else if (angular_vel > target_angular_vel) {
+		// Must decrease angular speed
+		if (CONFIG_angular_neg_accel_limit == 0.0 
+			|| angular_vel - target_angular_vel < CONFIG_angular_neg_accel_limit)
+			angular_vel = target_angular_vel;
+		else
+			 angular_vel -= CONFIG_angular_neg_accel_limit; 
+	}
+
+    float forward_displacement = dt * linear_vel * CONFIG_linear_odom_scale;
+	float yaw_displacement = dt * angular_vel * CONFIG_angular_odom_scale;
+	
+    yaw_rate = dt * angular_vel;
+    
+    // Integrate the displacements over time
+    // Update accumulated odometries and calculate the x and y components of velocity
+    odometry_w = yaw_displacement;
+    float delta_odometry_x = forward_displacement * std::cos(odometry_w);
+    float delta_odometry_y = forward_displacement * std::sin(odometry_w);
+    vel_x = delta_odometry_x;
+    vel_y = delta_odometry_y;
+    odometry_x += delta_odometry_x;
+    odometry_y += delta_odometry_y;
+	
+    // Create a Quaternion from the yaw displacement
+    quat = tf::createQuaternionMsgFromYaw(yaw_displacement);
+    last_time = current_time;
+ 
+    Eigen::Vector2f vec_linear_vel(vel_x, vel_y);
+    vel_.translation = vec_linear_vel;
+	vel_.angle = yaw_rate;
+
+    pose_.translation += Eigen::Rotation2Df(pose_.angle) * vel_.translation;
+    pose_.angle += yaw_displacement; 
+
+	PublishOdom(dt);
+}
+
+void DiffDriveModel::PublishOdom(const float dt) {
+    odom_msg.header.stamp = last_time;
+    odom_msg.pose.pose.position.x = odometry_x;
+    odom_msg.pose.pose.position.y = odometry_y;
+    odom_msg.pose.pose.position.z = 0.0;
+    odom_msg.pose.pose.orientation = quat;
```

Comment:
I'm confused about these assignments here. `quat` is created from `yaw_displacement` which is the change in angle in the last frame. It seems to me that the pose in the odom message should be the same as `pose_`. They are both odom->base_footprint.

Reply:
Just saw your latest commit. I don't believe `odometry_` == `pose_` mimics the behavior of the real robot, right? Odom tracks the relative movement of the robot, while my interpretation of `pose_` is that it tracks the position of the robot on the map.
It seems odd to me that odometry would contain the ground truth pose of the robot.

Reply:
It's usually not true, but the simulator seems to make that assumption. `pose_` is published as the odom->base_footprint transform [here](https://github.com/ut-amrl/f1tenth_simulator/blob/adf055bd04298da154f8125d74d68a689c47d28c/src/simulator/simulator.cpp#L297) and [here](https://github.com/ut-amrl/f1tenth_simulator/blob/adf055bd04298da154f8125d74d68a689c47d28c/src/simulator/simulator.cpp#L379). So for now `pose_` should match with what we publish to \odom. We can make another variable to track map->odom or map->base_footprint.

By the way, I believe the twist in the odometry message should be in the base_footprint frame. It's changed in that commit too.

Reply:
Hmm, I see. Will having odom specified that way (in the two links above) make AMCL's localization unrealistically accurate? Or does AMCL not use odometry information in that way?

Yes, I think you're right about the frame change, nice catch!

Reply:
AMCL uses odom->base_footprint and publishes map->odom, so it should be very accurate unless we fake noise in the odometry.